tests/unit/test_messageDialog.py

```diff
@@ -0,0 +1,431 @@
+# See the file COPYING for more details.
+# Copyright (C) 2024 NV Access Limited
+
+"""Unit tests for the message dialog API."""
+
+import unittest
+from unittest.mock import MagicMock, patch
+
+import wx
+from gui.messageDialog import (
+	DefaultButton,
+	DefaultButtonSet,
+	MessageDialog,
+	Button,
+	EscapeCode,
+	ReturnCode,
+	DialogType,
+	_flattenButtons,
+)
+
+
+NO_CALLBACK = (EscapeCode.NONE, None)
+
+
+def dummyCallback1(*a):
+	pass
+
+
+def dummyCallback2(*a):
+	pass
+
+
+class MDTestBase(unittest.TestCase):
+	"""Base class for test cases testing MessageDialog. Handles wx initialisation."""
+
+	def setUp(self) -> None:
+		self.app = wx.App()
+		self.dialog = MessageDialog(None, "Test dialog", buttons=None)
+
+
+@patch.object(wx.ArtProvider, "GetIconBundle")
+class Test_MessageDialog_Icons(MDTestBase):
+	"""Test that message dialog icons are set correctly."""
+
+	def test_setIcon_with_type_with_icon(self, mocked_GetIconBundle: MagicMock):
+		"""Test that setting the dialog's icons has an effect when the dialog's type has icons."""
+		mocked_GetIconBundle.return_value = wx.IconBundle()
+		type = DialogType.ERROR
+		self.dialog._setIcon(type)
+		mocked_GetIconBundle.assert_called_once()
+
+	def test_setIcon_with_type_without_icon(self, mocked_GetIconBundle: MagicMock):
+		"""Test that setting the dialog's icons doesn't have an effect when the dialog's type doesn't have icons."""
+		type = DialogType.STANDARD
+		self.dialog._setIcon(type)
+		mocked_GetIconBundle.assert_not_called()
+
+
+@patch("winsound.MessageBeep")
+class Test_MessageDialog_Sounds(MDTestBase):
+	"""Test that message dialog sounds are set and played correctly."""
+
+	def test_playSound_with_type_with_Sound(self, mocked_MessageBeep: MagicMock):
+		"""Test that sounds are played for message dialogs whose type has an associated sound."""
+		type = DialogType.ERROR
+		self.dialog._setSound(type)
+		self.dialog._playSound()
+		mocked_MessageBeep.assert_called_once()
+
+	def test_playSound_with_type_without_Sound(self, mocked_MessageBeep: MagicMock):
+		"""Test that no sounds are played for message dialogs whose type has an associated sound."""
+		type = DialogType.STANDARD
+		self.dialog._setSound(type)
+		self.dialog._playSound()
+		mocked_MessageBeep.assert_not_called()
+
+
+class Test_MessageDialog_Buttons(MDTestBase):
+	def test_addOkButton(self):
+		"""Test adding an OK button to the dialog."""
+		self.dialog.addOkButton()
+		with self.subTest("Check button types"):
+			self.assertIsInstance(self.dialog.FindWindowById(wx.ID_OK), wx.Button)
+		with self.subTest("Test in main buttons"):
+			self.assertEqual(self.dialog.GetMainButtonIds(), [wx.ID_OK])
+		with self.subTest("Test has default action."):
+			self.assertTrue(self.dialog.hasDefaultAction)
+		with self.subTest("Test default action assignment."):
+			id, callback = self.dialog._getDefaultAction()
+			self.assertEqual(id, ReturnCode.OK)
+			self.assertIsNotNone(callback)
```

Comment:
could these tests for adding buttons be abstracted / made DRY?

Reply:
Not easily with unittest alone. The [parameterized package](https://github.com/wolever/parameterized/) package seems to be a good drop-in means of adding parameterized tests to unittest though. This can also be done with a lot of fiddling with metaclasses, but I don't want to add code that needs to be tested to tests.
https://github.com/nvaccess/nvda/blob/bd63423cb9486c0829138fe65e8608fb225ff84f/tests/unit/test_messageDialog.py#L90-L153